benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90

```diff
@@ -69,860 +71,808 @@ SUBROUTINE invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, &
-    END SUBROUTINE momentum_u_code_set_args
+        use kind_params_mod, only: go_wp
+        USE clfortran
+        USE fortcl, ONLY: get_cmd_queues
+        real(go_wp), dimension(:,:), intent(in), target :: from
+        type(c_ptr), intent(in) :: to
+        integer, intent(in) :: startx, starty, nx, ny
+        logical, intent(in) :: blocking
+        integer(c_intptr_t) :: cl_mem
+        INTEGER(c_size_t) :: size_in_bytes, offset_in_bytes
+        INTEGER(c_intptr_t), pointer :: cmd_queues(:)
+        integer :: ierr
+        ! Copy complete ny rows (regardless of nx)
+        size_in_bytes = int(size(from, 1) * ny, 8) * c_sizeof(from(1,1))
+        offset_in_bytes = int(size(from, 1) * (starty - 1)) * c_sizeof(from(1,1))
+        cl_mem = transfer(to, cl_mem)
+        cmd_queues => get_cmd_queues()
+        ierr = clEnqueueWriteBuffer(cmd_queues(1), cl_mem, &
+            CL_TRUE, offset_in_bytes, size_in_bytes, C_LOC(from), 0, &
+            C_NULL_PTR, C_NULL_PTR)
+        CALL check_status('clEnqueueWriteBuffer', ierr)
+    end subroutine write_opencl
 
-    SUBROUTINE momentum_v_code_set_args(kernel_obj, va, un, vn, hu, hv, ht, &
-            ssha_v, sshn_t, sshn_u, sshn_v, tmask, dx_v, dx_t, &
-            dy_u, dy_v, dy_t, area_v, gphiv, omega, d2r, g, rdt, cbfr, visc)
-        USE clfortran, ONLY: clSetKernelArg
-        USE iso_c_binding, ONLY: c_sizeof, c_loc, c_intptr_t
+    subroutine initialise_device_grid(field)
+        USE fortcl, ONLY: create_rw_buffer
+        type(r2d_field), intent(inout), target :: field
+        integer(kind=c_size_t) size_in_bytes
+        IF (field%grid%tmask_device == 0) THEN
+            ! Create integer grid fields
+            size_in_bytes = int(field%grid%nx * field%grid%ny, 8) * &
+                            c_sizeof(field%grid%tmask(1,1))
+            field%grid%tmask_device = create_rw_buffer(size_in_bytes)
+
+            ! Create real grid buffers
+            size_in_bytes = int(field%grid%nx * field%grid%ny, 8) * &
+                            c_sizeof(field%grid%area_t(1,1))
+            field%grid%area_t_device = create_rw_buffer(size_in_bytes)
+            field%grid%area_u_device = create_rw_buffer(size_in_bytes)
+            field%grid%area_v_device = create_rw_buffer(size_in_bytes)
+            field%grid%dx_u_device = create_rw_buffer(size_in_bytes)
+            field%grid%dx_v_device = create_rw_buffer(size_in_bytes)
+            field%grid%dx_t_device = create_rw_buffer(size_in_bytes)
+            field%grid%dy_u_device = create_rw_buffer(size_in_bytes)
+            field%grid%dy_v_device = create_rw_buffer(size_in_bytes)
+            field%grid%dy_t_device = create_rw_buffer(size_in_bytes)
+            field%grid%gphiu_device = create_rw_buffer(size_in_bytes)
+            field%grid%gphiv_device = create_rw_buffer(size_in_bytes)
+        END IF
+    end subroutine initialise_device_grid
+
+    subroutine write_device_grid(field)
+        USE fortcl, ONLY: get_cmd_queues
```

Comment:
Please add a description of this routine.

Reply:
Done